### PR TITLE
Shayan/FEQ-911/github-actions-implementation

### PR DESCRIPTION
**Pull request by shayan-deriv**
we are migrating from `circle-ci` to `github actions`.
I moved reusable parts to actions which are:

- build
- invalidate_npm_cache
- npm_install_from_cache
- publish_to_cf_pages_staging
- publish_to_cf_pages_production
- send_slack_notifications
- setup_node
- test
- versioning

and also added two workflows : 

- release_staging
- release_production
- test

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    branches:
+      - '*'
+      - '!master'
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    container:
+      image: node:18.16
+    steps:
+    # - uses: actions/checkout@v4.1.0
```
Comment (shayan-deriv) [inline .github/workflows/test.yml:13]: do we need this?

File: .github/actions/setup_node/action.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  node-version:
+    description: 'The version of Node.js to use'
+    required: true
+runs:
+  using: composite
+  steps:
+  - name: Use Node.js ${{ inputs.node-version}}
+    uses: actions/setup-node@v3
+    with:
+      node-version: ${{ inputs.node-version}}
```
Comment (jim-deriv) [inline .github/actions/setup_node/action.yml:13]: @shayan-deriv,  please address the `EOL`

File: .github/actions/test/action.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+name: run_tests
+description: Run tests
+runs:
+    using: composite
+    steps:
+    - name: run tests
+      run: |
+        for i in {1..3}; do
+          npm test && break
+        done
```
Comment (jim-deriv) [inline .github/actions/test/action.yml:10]:
```suggestion
        for i in {1..3}; do
          npm test && break
        done
```
@shayan-deriv, could you please clarify on why the `for` `loop` is necessary? 🤔

File: .github/actions/setup_node/action.yml (hunk 2 above)
Comment (shayan-deriv) [reply .github/actions/setup_node/action.yml]: fixed 🙏

File: .github/actions/versioning/action.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (jim-deriv) [inline .github/actions/versioning/action.yml:1]:
```suggestion
```

File: .github/actions/test/action.yml (hunk 3 above)
Comment (shayan-deriv) [reply .github/actions/test/action.yml:10]: it doesn't make sense for me either. I checked with the binary-bot team . will remove it

File: .github/workflows/release_staging.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                  GD_API_KEY: ${{ secrets.GD_API_KEY }}
+                  GD_APP_ID: ${{ secrets.GD_APP_ID }}
+                  GD_CLIENT_ID: ${{ secrets.GD_CLIENT_ID }}
+                  TRACKJS_TOKEN: ${{ secrets.TRACKJS_TOKEN }}
+            - name: Test
+              uses: './.github/actions/test'
+            - name: Versioning
+              uses: './.github/actions/versioning'
+              with:
+                  RELEASE_TYPE: staging
```
Comment (jim-deriv) [inline .github/workflows/release_staging.yml:31]:
```suggestion
```

File: .github/workflows/test.yml (hunk 1 above)
Comment (jim-deriv) [reply .github/workflows/test.yml]: @shayan-deriv, Yeah , I agree with you. It doesn't seem necessary to me either

File: .github/actions/publish_to_cf_pages_production/action.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  CLOUDFLARE_ACCOUNT_ID:
+    description: "Cloudflare account id"
+    required: true
+  CLOUDFLARE_API_TOKEN:
+    description: "Cloudflare token"
+    required: true
+runs:
+  using: composite
+  steps:
+    - name: Publish to cloudflare pages (production)
```
Comment (ali-hosseini-deriv) [inline .github/actions/publish_to_cf_pages_production/action.yml:13]:
```suggestion
    - name: Publish to cloudflare pages (production)
      env:
          CLOUDFLARE_ACCOUNT_ID: ${{ inputs.CLOUDFLARE_ACCOUNT_ID }}
          CLOUDFLARE_API_TOKEN: ${{ inputs.CLOUDFLARE_API_TOKEN }}       
```
you need to pass the inputs as env to be pickable by wrangler

File: .github/actions/publish_to_cf_pages_staging/action.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  CLOUDFLARE_ACCOUNT_ID:
+    description: "Cloudflare account id"
+    required: true
+  CLOUDFLARE_API_TOKEN:
+    description: "Cloudflare token"
+    required: true
+runs:
+  using: composite
+  steps:
+    - name: Publish to cloudflare pages (staging)
```
Comment (ali-hosseini-deriv) [inline .github/actions/publish_to_cf_pages_staging/action.yml:13]: ditto

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    test:
+        runs-on: ubuntu-latest
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4.1.0
+            - name: Setup Node
+              uses: './.github/actions/setup_node'
+              with:
+                  node-version: '18.x'
+            - name: Git Cleanup
```
Comment (ali-hosseini-deriv) [inline .github/workflows/test.yml:16]: no need

File: .github/actions/setup_node/action.yml (hunk 2 above)
Comment (ali-hosseini-deriv) [inline .github/actions/setup_node/action.yml:13]:
```suggestion
      node-version: ${{ inputs.node-version }}
```

**Review by ali-hosseini-deriv — CHANGES_REQUESTED** [review]

File: .github/actions/publish_to_cf_pages_staging/action.yml (hunk 7 above)
Comment (shayan-deriv) [reply .github/actions/publish_to_cf_pages_staging/action.yml:13]: done

File: .github/actions/publish_to_cf_pages_production/action.yml (hunk 6 above)
Comment (shayan-deriv) [reply .github/actions/publish_to_cf_pages_production/action.yml:13]: done

File: .github/workflows/release_production.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            - name: Setup Node
+              uses: './.github/actions/setup_node'
+              with:
+                  node-version: '18.x'
+            - name: Install Dependencies
+              uses: './.github/actions/npm_install_from_cache'
+            - name: Build
+              uses: './.github/actions/build'
+              with:
+                  GD_API_KEY: ${{ secrets.GD_API_KEY }}
```
Comment (ali-hosseini-deriv) [inline .github/workflows/release_production.yml:24]: @vinu-deriv @rupato-deriv do we have different values for these env variables?

File: .github/workflows/release_production.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            - name: Extract version
+              id: extract_version
+              run: echo "RELEASE_VERSION=${VERSION}" >> $GITHUB_OUTPUT
+            - name: Deploy to Cloudflare Pages
+              uses: './.github/actions/publish_to_cf_pages_production'
+              with:
+                  CLOUDFLARE_ACCOUNT_ID: ${{ secrets.CLOUDFLARE_ACCOUNT_ID }}
+                  CLOUDFLARE_API_TOKEN: ${{ secrets.CLOUDFLARE_API_TOKEN }}
+    send_slack_notification:
+        if: always()
```
Comment (amir-deriv) [inline .github/workflows/release_production.yml:43]: if it's always going to run, do we need this condition here ?

File: .github/workflows/release_production.yml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+              uses: actions/checkout@v4
+            - name: Conclusion
+              uses: technote-space/workflow-conclusion-action@v3
+            - name: Send Slack Notification
+              uses: './.github/actions/send_slack_notifications'
+              with:
+                  SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+                  status: ${{ env.WORKFLOW_CONCLUSION }}
+                  release_type: Production
+                  version: ${{ needs.release_production.outputs.RELEASE_VERSION}}
```
Comment (amir-deriv) [inline .github/workflows/release_production.yml:57]: fix EOF

File: .github/workflows/release_staging.yml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+                  GD_APP_ID: ${{ secrets.GD_APP_ID }}
+                  GD_CLIENT_ID: ${{ secrets.GD_CLIENT_ID }}
+                  TRACKJS_TOKEN: ${{ secrets.TRACKJS_TOKEN }}
+            - name: Test
+              uses: './.github/actions/test'
+            - name: Deploy to Cloudflare Pages
+              uses: './.github/actions/publish_to_cf_pages_staging'
+              with:
+                  CLOUDFLARE_ACCOUNT_ID: ${{ secrets.CLOUDFLARE_ACCOUNT_ID }}
+                  CLOUDFLARE_API_TOKEN: ${{ secrets.CLOUDFLARE_API_TOKEN }}
```
Comment (amir-deriv) [inline .github/workflows/release_staging.yml:32]: fix EOF

File: .github/actions/setup_node/action.yml [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  node-version:
+    description: 'The version of Node.js to use'
+    required: true
+runs:
+  using: composite
+  steps:
+  - name: Use Node.js ${{ inputs.node-version }}
+    uses: actions/setup-node@v3
+    with:
+      node-version: ${{ inputs.node-version }}
```
Comment (amir-deriv) [inline .github/actions/setup_node/action.yml:13]: fix EOF

File: .github/actions/send_slack_notifications/action.yml [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    - name: Send Slack Notification on Failure
+      if: inputs.status == 'failure'
+      run: |
+        curl -X POST -H 'Content-type: application/json' \
+        --data '{
+          "text": "${{ inputs.RELEASE_TYPE }} Release failed for deriv bot with version ${{ inputs.VERSION }}"
+        }' \
+        ${{ inputs.SLACK_WEBHOOK_URL }}
+      shell: bash
```
Comment (amir-deriv) [inline .github/actions/send_slack_notifications/action.yml:39]: fix EOF

File: .github/actions/publish_to_cf_pages_staging/action.yml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+      env:
+          CLOUDFLARE_ACCOUNT_ID: ${{ inputs.CLOUDFLARE_ACCOUNT_ID }}
+          CLOUDFLARE_API_TOKEN: ${{ inputs.CLOUDFLARE_API_TOKEN }}
+      run: |-
+        cd www
+        npx wrangler pages publish . --project-name=deriv-binary-bot-pages --branch=main
+        echo "New staging website - http://staging.cf-pages-deriv-binary-bot.deriv.com"
+      shell: bash
```
Comment (amir-deriv) [inline .github/actions/publish_to_cf_pages_staging/action.yml:21]: fix EOF

File: .github/actions/publish_to_cf_pages_production/action.yml [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+      env:
+          CLOUDFLARE_ACCOUNT_ID: ${{ inputs.CLOUDFLARE_ACCOUNT_ID }}
+          CLOUDFLARE_API_TOKEN: ${{ inputs.CLOUDFLARE_API_TOKEN }}
+      run: |-
+        cd www
+        npx wrangler pages publish . --project-name=deriv-binary-bot-pages --branch=main
+        echo "New website - http://cf-pages-deriv-binary-bot.deriv.com"
+      shell: bash
```
Comment (amir-deriv) [inline .github/actions/publish_to_cf_pages_production/action.yml:21]: fix EOF

File: .github/workflows/release_production.yml [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            - name: Checkout
+              uses: actions/checkout@v4
+            - name: Setup Node
+              uses: './.github/actions/setup_node'
+              with:
+                  node-version: '18.x'
+            - name: Install Dependencies
+              uses: './.github/actions/npm_install_from_cache'
+            - name: Build
+              uses: './.github/actions/build'
```
Comment (amir-deriv) [inline .github/workflows/release_production.yml:22]: indentation looks inconsistent with the other `.yml` files

File: .github/workflows/release_production.yml (hunk 10 above)
Comment (shayan-deriv) [reply .github/workflows/release_production.yml:43]: we still need it. by adding it , the job will run even if the previous job fails. 
https://stackoverflow.com/questions/58858429/how-to-run-a-github-actions-step-even-if-the-previous-step-fails-while-still-f

File: .github/workflows/release_production.yml (hunk 17 above)
Comment (shayan-deriv) [reply .github/workflows/release_production.yml]: fixed

**Review by jim-deriv — APPROVED** [review]

**Comment by rupato-deriv** [comment]
@ali-hosseini-deriv @shayan-deriv can you please check the node_version because last time when we were updating we had set it to 18 and the release was failing on staging so we had downgraded it to 16 because of cjs issue other than this all looks good 👍

File: .github/workflows/release_production.yml (hunk 9 above)
Comment (ali-hosseini-deriv) [reply .github/workflows/release_production.yml]: thanks @rupato-deriv 
@shayan-deriv then we have to setup separate `environment` attribute for staging and production workflows. 
please check [this](https://github.com/binary-com/deriv-app/blob/master/.github/workflows/release_production.yml#L9) as reference

Comment (shayan-deriv) [reply .github/workflows/release_production.yml]: it's added . thanks guys

**Pull request merged**